atat/domain/csp/cloud/azure_cloud_provider.py

```
@@ -1582,3 +1569,148 @@ def get_calculator_url(self):
+            )
+        except UnknownServerException as exc:
+            if exc.status_code == "403":
+                app.logger.warning(
+                    (
+                        "Tenant admin of tenant %s unable to list role assignments."
+                        "This could indicate that the tenant admin does not have a User"
+                        "Access Administrator role assignment."
+                    ),
+                    tenant_id,
+                )
+                return True
+            else:
+                raise exc
+
+        assignment = self.filter_role_assignments(role_assignments, definition_id)
+        if not assignment:
+            app.logger.warning(
+                "User Access Administrator role assignment not found for user %s in tenant %s",
+                tenant_admin_user_object_id,
+                tenant_id,
+            )
+            return True
+        return bool(self.remove_role_assignment(token, assignment["name"]))
+
+    @contextmanager
+    def get_elevated_access_token(self, tenant_id, user_object_id):
+        tenant_admin_token = self._get_tenant_admin_token(
+            tenant_id, self.sdk.cloud.endpoints.resource_manager + "/.default"
+        )
+        elevated_token = self._elevate_tenant_admin_access(tenant_admin_token)
+        try:
+            yield elevated_token
+        finally:
+            for _ in range(3):
+                try:
+                    self.remove_tenant_admin_elevated_access(tenant_id, user_object_id)
+                    return
+                except self.sdk.requests.exceptions.RequestException:
+                    time.sleep(5)
+
+            app.logger.error(
+                "Could not remove User Access Administrator for user %s in tenant %s",
+                tenant_id,
+                user_object_id,
+            )
+            raise self.sdk.requests.exceptions.RequestException(
+                "Error removing elevated access"
+            )
```

At this point, the tenant admin has successfully elevated their access and has attempted a portfolio provisioning step. Whether or not the provisioning step has succeeded, we try to remove the elevated access. 

We try the call to remove the elevated access within the exit block of the context manager 3 times to account for network errors. If the call still doesn't succeed, we want to raise an exception that isn't a `GeneralCSPException` so that Celery won't retry the job. The portfolio will remain in a failed state and will require manual intervention to continue provisioning resources and remove elevated access.

---

I've since removed this retry functionality. It appears that `return` statement in the `for` loop was messing with the function that was being called within the context manager. The function would return `None` instead of the intended return value.

While the [`@contextmanger`](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) decorator is a little more convenient to use, the complexity that we're managing here might mean that we should upgrade to a context manager [class ](https://docs.python.org/3/library/stdtypes.html#context-manager-types). I don't know if that should happen in this PR or follow on.